src/Event/ApiGateway/ApiGatewayHandler.php

```diff
@@ -0,0 +1,20 @@
+<?php declare(strict_types=1);
+
+namespace Bref\Event\ApiGateway;
+
+use Bref\Context\Context;
+use Bref\Event\Handler;
+
+/**
+ * Handles ApiGateway events.
+ */
+abstract class ApiGatewayHandler implements Handler
+{
+    abstract public function handleWebsocket(WebsocketEvent $event, Context $context): WebsocketResponse;
+
+    /** {@inheritDoc} */
+    public function handle($event, Context $context)
+    {
+        return $this->handleWebsocket(new WebsocketEvent($event), $context)->toApiGatewayFormat();
```

Comment:
Is a websocket response only a status code?

Because I'm wondering if we need the `WebsocketResponse` class at all? (we could return the status code a `int`).

One argument for keeping the class would be if there are other fields (not implemented yet), or if there will be (is it the case?).

Reply:
@mnapoli You've got a point. Initially I thought that response looked similar to HTTP responses and thus took inspiration in how it handled it's response. What I found how-ever was that forwarding a 'body' in the response broke it and connection received an error. Looking at the documentation of Websockets further I've found that it does support headers: https://docs.aws.amazon.com/apigateway/latest/developerguide/websocket-connect-route-subprotocol.html

Maybe to simplify things we should have the response accept both a return type of int and WebsocketResponse. This way we could add in support for headers in WebsocketResponse and just have the handler wrap int returns in the response class for ApiGateway formatting?This would means that for 9/10 scenarios where you just want to return a status code you just return the int, but still leaves room for future additions to the ApiGateway Websocket API.

Reply:
This obviously also depends on the rest of the code base. If it's more in tune doing string values - then I'll just change it. Better to have same style for each implementation.

Reply:
I've done some further investigation and it appears only `Sec-WebSocket-Protocol` can be specified in the headers object. I'm unable to find much else on it besides the aws docs already forwarded and https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api

Updating PR to support int|WebsocketResponse both as return type. Where as WebsocketResponse second constructor argument would be the supported protocols.

Reply:
Hi, really sorry, the fact that the response is actually an object with possible additional properties in JS is very helpful: keeping `WebsocketResponse` only makes a lot of sense then (and it's future-proof as well).

Could you drop the `int` then 😅 ? (the less choice there is the better)

Reply:
Wait so I had a look around, I'm not sure what to think TBH. The examples I've found online look a lot like a HTTP response ([e.g. this example by AWS](https://github.com/aws-samples/simple-websockets-chat-app/blob/e4566f8a30986e53e90024c92a41ea6f5f863e18/sendmessage/app.js#L39-L45)).

But I'm not sure if that response format is required by API Gateway (and not used by websockets itself), or if a websocket response is actually a HTTP response (which I don't think it is but I don't really know much about websockets).

So I don't know if we should have a custom class for that response, or use a HTTP response class, or do something else 🤔

Reply:
@mnapoli Hm - that's interesting. I was unable to find anything officially in the documentation documenting the 'body' index. & I'm also not familiar with websockets having the ability to return data in the connect, message or & disconnect events. Maybe this is just their way re-using ApiGateway's existing & powerful API? I'll do some further investigation on this.

Initially I tried supplying a body, but what I experienced was that ApiGateway websocket connect got rejected to due a syntax error. If websocket support statusCode, headers & body just like normal HTTP responses for ApiGateway - then we could as you wrote reuse HttpResponse for websockets. But that being said, I really don't see where 'body' comes into play.

I will do some more investigation on this and come back with my findings.

Reply:
@mnapoli  Apparently I was mistaken regarding `body` index in response. It's perfectly acceptable and ApiGateway doesn't react to it at all. I've gone ahead and made the change so it now expect a HttpResponse instead of the new WebsocketResponse.

Looked at the [RFC](https://tools.ietf.org/html/rfc6455#page-6) and I see no mention of any possibility of directly giving a response other than a status code for either events. Even sending a 500 status code error on a message event does nothing in the frontend (no exception or 'onerror' trigger). The only event that allows for anything other than a status code to be put as response appears to be the connect event. Within connect event you can specify a few headers (also mentioned in https://tools.ietf.org/html/rfc6455#page-6).

It's very unclear to me why Amazon has even put a `body` index in their example. When looking from a RFC standpoint there is absolutely no way of sending anything back in either event matching a response body. In addition to this it doesn't appear to do anything whether you're responding with a body or not.

I'm conflicted. Utilizing the HttpResponse class does make sense as it appears to use the exact same response structure as for Http requests. But - HttpResponse expects first constructor argument to be a string (not allowing null). This means you'll have to provide a empty string as a body or anything really just to signal a 200 success. At the same time - supporting HttpResponse makes the event handling more in tune with AWS's own examples 😕 

-

Update:

Asked AWS regarding this and got the following:
```
Them: So generally speaking, it is more used for Data Transformations in the Integration Response. If for example you're using the same Lambda Function to provide responses on both a REST API and a Websocket API, you can use the Data Mapping templates to take the body being returned from the function and map it into a part of the response supported by the protocol.

Me: Ah. So it's a way of enabling the same lambda function to respond to regular Http/Rest requests from ApiGateway?

Them: That is one example use case, I personally don't have a wide range of experience with the Websocket API Gateways, but I can take you through the Data Transformations to help you better understand the implementation.

Me: That would be very helpful for understanding it's possible use-cases. Do you have an example of this? Just to get the idea?

Them: So this page of the docs has a couple of examples for Integration Request mappings, but the same approach can be used to remap data in an Integration Response.

https://docs.aws.amazon.com/apigateway/latest/developerguide/websocket-api-data-mapping.html

When dealing with REST APIs you can configure these within the AWS Console, however with a Websocket API you have to use SDK/CLI to implement

And follow this, plus the Data mapping and WebSocket mapping template reference pages underneath, you'll be able to see that it is quite powerful, and will allow you to re-use existing Lambdas/HTTP Servers and offload the work of re-working data for Websockets onto API gateway instead of having to build another method.
```

So all in all i suppose they're one and the same (Rest, Http & Websocket). Which means it would make sense to utilize the same HttpResponse, right?

Reply:
@mnapoli Let me know if there is anything i can do on this. I know there is a lot of dialog / discussion on this. I got the time to adjust any changes you'd like on this if need be.

Reply:
Very sorry about the delay, I was unsure about merging something I'm unsure about.

Look, we have one life, and worst case scenario we got the return type wrong: we can always remove it later and support 2 kind of returned value. v1.0 is also around the corner, so there's that. The rest of the codebase is not impacted.

Let's do it!

Reply:
@mnapoli No worries at all. Appreciate it!